sql/distinct.go

```diff
@@ -0,0 +1,155 @@
+
+package sql
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// distinct constructs a distinctNode.
+func (*planner) distinct(n *parser.Select) *distinctNode {
+	if n.Distinct {
+		d := distinctNode{}
+		d.prefixSeen = []byte{}
+		d.suffixSeen = make(map[string]struct{})
+		return &d
+	}
+	return nil
+}
+
+type distinctNode struct {
+	planNode
+	// All the columns that are part of the Sort. Set to nil if no-sort, or
+	// sort used an expression that was not part of the requested column set.
+	columnsInOrder []bool
+	// encoding of the columnsInOrder columns for the previous row.
+	prefixSeen []byte
+	// encoding of the non-columnInOrder columns for rows sharing the same
+	// prefixSeen value.
+	suffixSeen map[string]struct{}
+	err        error
+}
+
+// wrap the supplied planNode with the distinctNode if distinct is required.
+func (n *distinctNode) wrap(plan planNode) planNode {
+	if n == nil {
+		return plan
+	}
+	n.planNode = plan
+	if len(plan.Ordering()) == 0 {
+		return n
+	}
+	n.columnsInOrder = make([]bool, len(plan.Columns()))
+	for _, p := range plan.Ordering() {
+		if p < 0 {
+			p = -p
+		}
+		if p <= len(n.columnsInOrder) {
+			n.columnsInOrder[p-1] = true
+		} else {
+			// Cannot use sort order. This happens when the
+			// columns used for sorting are not part of the output.
+			// e.g. SELECT a FROM t ORDER BY c.
+			n.columnsInOrder = nil
+			break
+		}
+	}
+	return n
+}
+
+func (n *distinctNode) Next() bool {
+	for n.planNode.Next() {
+		// Detect duplicates
+		prefix, suffix := n.encodeValues(n.Values())
+		if n.err != nil {
+			return false
+		}
+		if bytes.Equal(prefix, n.prefixSeen) {
+			// The prefix of the row is the same as the last row; check
+			// to see if the suffix which is not ordered has been seen.
+			if suffix == nil {
+				// duplicate
+				continue
+			}
+			sKey := string(suffix)
+			if _, ok := n.suffixSeen[sKey]; ok {
+				// duplicate
+				continue
+			}
+			n.suffixSeen[sKey] = struct{}{}
+		} else {
+			// The prefix of the row which is ordered differs from the last row;
+			// reset our seen set.
+			if len(n.suffixSeen) > 0 {
+				n.suffixSeen = make(map[string]struct{})
+			}
+			n.prefixSeen = prefix
+			if suffix != nil {
+				n.suffixSeen[string(suffix)] = struct{}{}
+			}
+		}
+		return true
+	}
+	n.err = n.planNode.Err()
+	return false
+}
+
+func (n *distinctNode) Err() error {
+	return n.err
+}
+
+func (n *distinctNode) encodeValues(values parser.DTuple) ([]byte, []byte) {
+	var prefix, suffix []byte
+	for i, val := range values {
+		var err error
+		if n.columnsInOrder != nil && n.columnsInOrder[i] {
+			if prefix == nil {
+				prefix = make([]byte, 0, 100)
+			}
+			prefix, err = encodeTableKey(prefix, val)
+		} else {
+			if suffix == nil {
+				suffix = make([]byte, 0, 100)
+			}
+			suffix, err = encodeTableKey(suffix, val)
+		}
+		if err != nil {
+			n.err = util.Errorf("unable to encode value: %v", val)
```

Comment:
Why not just set `n.err = err`? Or even better, `prefix, n.err = encodeTableKey(...)` and `suffix, n.err = encodeTableKey(...)`. Note that you should exit out of this loop if `n.err != nil`.